test/unit/spec/twilioconnection.js

```diff
@@ -173,185 +234,294 @@ describe('TwilioConnection', function() {
+                          assert.equal(typeof hello.id, 'string');
+                          assert.equal(hello.timeout, twilioConnection._options.requestedHeartbeatTimeout);
+                          assert.equal(hello.type, 'hello');
+                          assert.equal(hello.cookie, 'foo');
+                          if (options.helloBody) {
+                            assert.equal(hello.body, options.helloBody);
+                          } else {
+                            assert.equal('body' in hello, false);
+                          }
+                        });
+                      });
+                    }
+                  }
+                });
+              });
+            });
           });
-          error = await promise;
-        });
 
-        it('should set the TwilioConnection\'s .state to "closed', () => {
-          assert.equal(twilioConnection.state, 'closed');
-        });
+          context('when a "welcome" message is not received within the "welcome" timeout', () => {
+            let error;
 
-        it('should call .close on the underlying WebSocket', () => {
-          sinon.assert.callCount(twilioConnection._ws.close, 1);
-        });
+            beforeEach(async () => {
+              error = await new Promise(resolve => {
+                twilioConnection.once('close', resolve);
+              });
+            });
 
-        it('should emit "close" on the TwilioConnection with the appropriate Error', () => {
-          assert(error instanceof Error);
-          assert(/^WebSocket Error 3002/.test(error.message));
-        });
-      });
+            it('should set the TwilioConnection\'s .state to "closed', () => {
+              assert.equal(twilioConnection.state, 'closed');
+            });
 
-      context('when a "welcome" message is not received within the "welcome" timeout', () => {
-        let error;
+            it('should call .close on the underlying WebSocket', () => {
+              sinon.assert.callCount(twilioConnection._ws.close, 1);
+            });
 
-        beforeEach(async () => {
-          error = await new Promise(resolve => {
-            twilioConnection.once('close', resolve);
+            it('should emit "close" on the TwilioConnection with the appropriate Error', () => {
+              assert(error instanceof Error);
+              assert(/^WebSocket Error 3000/.test(error.message));
+            });
           });
         });
-
-        it('should set the TwilioConnection\'s .state to "closed', () => {
-          assert.equal(twilioConnection.state, 'closed');
-        });
-
-        it('should call .close on the underlying WebSocket', () => {
-          sinon.assert.callCount(twilioConnection._ws.close, 1);
-        });
-
-        it('should emit "close" on the TwilioConnection with the appropriate Error', () => {
-          assert(error instanceof Error);
-          assert(/^WebSocket Error 3000/.test(error.message));
-        });
       });
```

Comment:
It looks like many of the individual cases are being tested, but it may be worthwhile to try the full busy call flow including more than one loop through early -> connecting -> wait -> ...

Reply:
@ceaglest We do have a test for `early -> connecting -> busy -> connecting -> open` as well as `early -> connecting -> busy -> early -> connecting -> open`.